### PR TITLE
@/__matmul__ now only performs matrix multiplication

### LINKED ISSUE #13766: @ (__matmul__) should fail if one argument is not a matrix
```
>>> A = Matrix([[1, 2], [3, 4]])
>>> B = Matrix([[2, 3], [1, 2]])
>>> A@B
Matrix([
[ 4,  7],
[10, 17]])
>>> 2@B
Matrix([
[4, 6],
[2, 4]])
```

Right now `@` (`__matmul__`) just copies `__mul__`, but it should actually only work if the multiplication is actually a matrix multiplication. 

This is also how NumPy works

```
>>> import numpy as np
>>> a = np.array([[1, 2], [3, 4]])
>>> 2*a
array([[2, 4],
       [6, 8]])
>>> 2@a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Scalar operands are not allowed, use '*' instead
```

### PR DESCRIPTION
**Pull request by abhigyank**
Add check to allow __matmul__ to perform only matrix multiplication and throws ValueError if
scalars are used as operators or operands.

#### References to other Issues or PRs

This PR will fix #13766 .

#### Brief description of what is fixed or changed
```
>>> B = Matrix([[2, 3], [1, 2]])
>>> 2@B
```
and
```
>>> B@2
```
now throws ValueError: Scalar operands are not allowed, use '*' instead

### CODE REVIEW

**Comment by abhigyank** [comment]
@asmeurer Have a look at it.

File: sympy/matrices/common.py [hunk 1]
```diff
@@ -1973,6 +1973,13 @@ def __div__(self, other):
 
     @call_highest_priority('__rmatmul__')
     def __matmul__(self, other):
+        other = _matrixify(other)
+        # matrix-like objects can have shapes.  This is our check.
+        # Non matrice operands will throw error
+        if not(hasattr(other, 'shape')):
+            raise ValueError("Scalar operands are not allowed, use '*' instead");
+            return
```
Comment (asmeurer) [inline sympy/matrices/common.py]: Remove the `return`.

File: sympy/matrices/common.py [hunk 2]
```diff
@@ -1973,6 +1973,13 @@ def __div__(self, other):
 
     @call_highest_priority('__rmatmul__')
     def __matmul__(self, other):
+        other = _matrixify(other)
+        # matrix-like objects can have shapes.  This is our check.
+        # Non matrice operands will throw error
+        if not(hasattr(other, 'shape')):
+            raise ValueError("Scalar operands are not allowed, use '*' instead");
```
Comment (asmeurer) [inline sympy/matrices/common.py]: Don't put semicolons at the end of lines.

**Comment by asmeurer** [comment]
There are two PRs fixing this, this one and https://github.com/sympy/sympy/pull/13772.

File: sympy/matrices/tests/test_commonmatrix.py [hunk 3]
```diff
@@ -674,6 +674,13 @@ def test_multiplication():
         assert c[1, 0] == 3*5
         assert c[1, 1] == 0
 
+def test_matmul():
+    a = Matrix([[1, 2], [3, 4]])
+    b = Matrix([[2, 3], [1, 2]])
+    raises(ValueError, lambda: a@2)
+    raises(ValueError, lambda: 2@a)
```
Comment (asmeurer) [inline sympy/matrices/tests/test_commonmatrix.py]: You will need to use eval for this test, and make it so that it skips in Python versions less than 3.5.

Comment (asmeurer) [reply sympy/matrices/tests/test_commonmatrix.py]: Or you can just test `__matmul__`.

**Comment by asmeurer** [comment]
Both PRs implement the fix in the same way, and both have some minor issues. @abhigyank, 	Subhash-Saurabh, how would you like to proceed here?

**Comment by Mohit17067** [comment]
I have no right to do anything here but there was a third person also who said before these two guys that he is going to work on this issue ( issue #13766 ). I think you guys should give him a chance also else it will be difficult for newcomers to solve issues. No offense to the two who solved it. The decision is all yours.

**Comment by gxyd** [comment]
I think @ArighnaIITG is already working on a few PR's. BTW there are few other issues that might need your love, for ex. https://github.com/sympy/sympy/issues/11345 is also based on matrices and is easy to fix, one of you can try that.

**Comment by abhigyank** [comment]
This is my first issue at such a big project and wasn't aware of other PRs. Its upto you to accept any PR. No issues there. 
@asmeurer 
I have pushed a commit making your recommended changes. Please look at it.

**Comment by Subhash-Saurabh** [comment]
@abhigyank You can continue with the work. In the mean time I will work on some other issues .

**Comment by asmeurer** [comment]
https://github.com/sympy/sympy/pull/13772#issuecomment-352860133 makes a good point. We should probably return NotImplemented instead of raising ValueError. It will give a less informative error message, but it will allow other objects to define `__matmul__` behavior on SymPy matrices easier.

**Comment by abhigyank** [comment]
I think NotImplemented error isn't appropriate as we actually had that implemented already, and now are removing that functionality since ```__matmul__``` should work for scaler operands. Raising a TypeError is still ok, but is on similar lines with ValueError.

**Comment by asmeurer** [comment]
`NotImplemented` is not the same as `NotImplementedError`. See https://docs.python.org/3/library/constants.html#NotImplemented

**Comment by abhigyank** [comment]
Okay, I understand. It seems appropriate in this case then. Do you want me to implement it?On Dec 20, 2017 2:45 PM, Aaron Meurer <notifications@github.com> wrote:NotImplemented is not the same as NotImplementedError. See https://docs.python.org/3/library/constants.html#NotImplemented

**Comment by asmeurer** [comment]
Yes

**Comment by abhigyank** [comment]
@asmeurer Have a look now.

**Comment by abhigyank** [comment]
I don't understand from the logs why the Travis build failed. What's the issue?

**Comment by abhigyank** [comment]
I think it can be merged @asmeurer

**Comment by asmeurer** [comment]
Add a test for `a @ 2`/`a.__rmatmul__(2)` as well.

File: sympy/matrices/tests/test_commonmatrix.py [hunk 4]
```diff
@@ -674,6 +674,20 @@ def test_multiplication():
...
 
+def test_matmul():
+    a = Matrix([[1, 2], [3, 4]])
+
+    #Check a@2 case
+    assert a.__matmul__(2) == NotImplemented
+
+    #To check 2@a case
+    try:
+        eval('2 @ a')
```
Comment (asmeurer) [inline sympy/matrices/tests/test_commonmatrix.py]: Add a comment here that this is done this way because @ is only supported in Python 3.5+.

**Comment by abhigyank** [comment]
@asmeurer I have done and pushed. Have and look and merge maybe.

**Comment by ylemkimon** [comment]
I think a check for `len(shape) == 2` is needed.

**Comment by abhigyank** [comment]
That check is already present in __mul__ .

**Comment by ylemkimon** [comment]
But if `len(shape) != 2`, it will perform a scalar multiplication.

**Comment by abhigyank** [comment]
@ylemkimon in line 1976 ```other = _matrixify(other)``` , this passes the other through _matrixify() function, which already contains check for ```is_Matrix``` and ```len(shape)==2``` and then returns other accordingly. So what we receive has already passed through the given conditions and hence we can only check for its matrix feature. The other checks are redundant and not required again.

**Comment by ylemkimon** [comment]
https://github.com/sympy/sympy/blob/7d0231704a21c5dbe4740e1f354f736b5a313acc/sympy/matrices/common.py#L2265-L2274 In `_matrixify`, if `len(shape) != 2`, it will return original object(`other`) unchanged. And since it has `shape` attribute, it will be passed to `__mul__`. In `__mul__`, if not `is_Matrix` and not `is_MatrixLike`, i.e., `len(shape) != 2`, it will be passed to `_eval_scalar_mul`, which is a scalar multiplication.

**Comment by abhigyank** [comment]
@ylemkimon Okay, I get your point. So ``` if (getattr(mat, 'is_Matrix', False) and getattr(mat, 'is_Matrix', False)): return NotImplemented ``` should fix this right?

**Comment by ylemkimon** [comment]
Yes, but you seems to have missed `not`. So, it will be look like:
```python
other = _matrixify(other)
if not getattr(other, 'is_Matrix', False) or not getattr(other, 'is_Matrix', False)):
    return NotImplemented
```

**Comment by abhigyank** [comment]
@ylemkimon  @asmeurer  Have a look now, and merge if ready.

File: sympy/matrices/common.py [hunk 5]
```diff
@@ -1973,6 +1973,14 @@ def __div__(self, other):
 
     @call_highest_priority('__rmatmul__')
     def __matmul__(self, other):
+        other = _matrixify(other)
+        # matrix-like objects can have shapes.  This is our check.
+        # Non matrice operands will throw error
+        if not(hasattr(other, 'shape')):
+            return NotImplemented
```
Comment (ylemkimon) [inline sympy/matrices/common.py]: `shape` attribute check is no longer needed.

**Review by ylemkimon — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by abhigyank** [comment]
@ylemkimon Amended the previous commit to do your requested changes. Have a look now @ylemkimon  @asmeurer

**Review by ylemkimon — APPROVED** [review]

**Comment by ylemkimon** [comment]
Thank you for your contribution!

**Pull request merged**